algebras/algebra/src/test/scala/endpoints/algebra/server/EndpointsTestSuite.scala

```diff
@@ -99,10 +101,66 @@ trait EndpointsTestSuite[T <: endpoints.algebra.Endpoints] extends ServerTestBas
 
     }
 
+    "xmap query strings" should {
+
+      "xmap urls of locations" in {
+        
+
+        val locationQueryString =
+          (qs[Double]("lon") & qs[Double]("lat"))
+            .xmap[Location](
+              { case (lon, lat) => Location(lon, lat) },
+              location => (location.longitude, location.latitude))
+        val locationUrl = path /? locationQueryString
+        
+        decodeUrl(locationUrl)("/?lon=12.0&lat=32.9") shouldEqual Matched(Location(12.0, 32.9))
+        decodeUrl(locationUrl)("/?lon=-12.0&lat=32.9") shouldEqual Matched(Location(-12.0, 32.9))
+        decodeUrl(locationUrl)(s"/?lon=${Math.PI}&lat=-32.9") shouldEqual Matched(Location(Math.PI, -32.9))
+        decodeUrl(locationUrl)("/?lon=12&lat=32") shouldEqual Matched(Location(12, 32))
+        decodeUrl(locationUrl)("/?lat=32.0&lon=12.0") shouldEqual Matched(Location(12.0, 32.0))
+
+        decodeUrl(locationUrl)("/?lon=12,0&lat=32.0") shouldEqual Malformed
+        decodeUrl(locationUrl)("/?lon=a&lat=32") shouldEqual Malformed
+        decodeUrl(locationUrl)("/?let=12.0&lat=32.0") shouldEqual Malformed
+        decodeUrl(locationUrl)("/?lon=12.0") shouldEqual Malformed
+      }
+
+      "xmapPartial urls of blogids" in {
+        val blogIdQueryString: QueryString[BlogId] =
+          (qs[Option[UUID]]("uuid") & qs[Option[String]]("slug"))
+            .xmapPartial {
+              case (maybeUuid, maybeSlug) =>
+                maybeUuid.map[BlogId](BlogUuid).orElse(maybeSlug.map[BlogId](BlogSlug))
+            } {
+              case BlogUuid(uuid) => (Some(uuid), None)
+              case BlogSlug(slug) => (None, Some(slug))
+            }
+        
+        val testUUID: String = "f4b9defa-1ad8-453f-9a06-2683b8564b8d"
```

Comment:
Maybe it’d be easier to work with an `UUID` value from the beginning? (it would save some `UUID.fromString` calls)

~~~ scala
val testUUID: UUID = UUID.randomUUID()
~~~